statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/backpressure/ThresholdBackPressureValve.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.statefun.flink.core.backpressure;
+
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashMap;
+import java.util.Objects;
+import org.apache.flink.statefun.sdk.Address;
+
+/** A simple Threshold based {@link BackPressureValve}. */
+public final class ThresholdBackPressureValve implements BackPressureValve {
+  private final int maximumPendingAsynchronousOperations;
+
+  /**
+   * a set of address that had explicitly requested to stop processing any new inputs (via {@link
+   * AsyncWaiter#awaitAsyncOperationComplete()}. Note that this is a set implemented on top of a
+   * map, and the value (Boolean) has no meaning.
+   */
+  private final ObjectOpenHashMap<Address, Boolean> blockedAddressSet =
+      new ObjectOpenHashMap<>(1024);
+
+  private int pendingAsynchronousOperationsCount;
+
+  /**
+   * Constructs a ThresholdBackPressureValve.
+   *
+   * @param maximumPendingAsynchronousOperations the total allowed async operations to be inflight
+   *     per StreamTask, or {@code -1} to disable back pressure.
+   */
+  public ThresholdBackPressureValve(int maximumPendingAsynchronousOperations) {
+    this.maximumPendingAsynchronousOperations = maximumPendingAsynchronousOperations;
+  }
+
+  public boolean shouldBackPressure() {
+    return totalPendingAsyncOperationsAtCapacity() || hasBlockedAddress();
+  }
+
+  public void blockAddress(Address address) {
```

Comment:
What happens if a function blocks itself (via `AsyncWaiter#awaitAsyncOperationComplete()`), when there actually are no in-flight async operations for it?
Since in that case, the `notifyAsyncOperationCompleted` seems to never be invoked for that address, and therefore the blocked address will never be removed from `blockedAddresSet` ==> `hasBlockedAddress` will end up always being false?

Not sure if I'm missing something here, please correct me if I'm wrong!

Reply:
Hi @tzulitai, your understanding is correct, If a function would request to block without actually registering any async operations then it would remain blocked. 
This is an acceptable behaviour since this is our internal API to be used by `HttpFunction` and `GrpcFunction`.

If we would like it to be a part of the SDK then we would have to make sure that we track every async operation registered per each address.

Let me update the JavaDoc of the `AsyncWaiter` with your observation.

Reply:
I see, understood now thank you!